File: apps/api/app/audit.py
from __future__ import annotations

from sqlalchemy.orm import Session

from .core import utc_now
from .models import AuditLog


def record_audit(
    session: Session,
    *,
    action: str,
    status: str = "success",
    workspace_id: str | None = None,
    actor_id: str | None = None,
    entity_type: str | None = None,
    entity_id: str | None = None,
    meta: dict | None = None,
) -> AuditLog:
    audit = AuditLog(
        workspace_id=workspace_id,
        actor_id=actor_id,
        action=action,
        status=status,
        entity_type=entity_type,
        entity_id=entity_id,
        meta_json=meta,
        created_at=utc_now(),
    )
    session.add(audit)
    return audit

File: apps/api/src/modules/audit.ts
import type { Kysely } from 'kysely'
import type { Database } from '../db/schema.js'
import { jsonString } from '../db/database.js'
import { newId } from '../core/security.js'
import { utcNow } from '../core/time.js'

export async function recordAudit(
  db: Kysely<Database>,
  input: {
    workspaceId?: string | null
    actorId?: string | null
    action: string
    status?: string
    entityType?: string | null
    entityId?: string | null
    meta?: unknown
  },
) {
  await db
    .insertInto('audit_logs')
    .values({
      id: newId(),
      workspace_id: input.workspaceId ?? null,
      actor_id: input.actorId ?? null,
      action: input.action,
      status: input.status ?? 'success',
      entity_type: input.entityType ?? null,
      entity_id: input.entityId ?? null,
      meta_json: input.meta === undefined ? null : jsonString(input.meta),
      created_at: utcNow(),
    })
    .execute()
}
